### DISCUSSION TITLE
Debug TB Without a Simulation

### DISCUSSION THREAD

**Discussion by zruyam** [pyuvm/pyuvm#134] [General]
Hi,
Is it possible to debug a testbench without running a simulation?
Thanks!

**Comment by raysalemi** [comment]
It was in an early version of **pyuvm** (1.0) that used threads for parallelism.

However, coroutines were such a better solution for parallelism that building **pyuvm** on top ov **cocotb** made the most sense.  

That said, the simulation does not have to do much or even contain the DUT.  I've never tried it, but I'll bet one could write an entire transaction-level testbench without interacting with the simulator.  

But, yes, there must be a simulator running.
